File: convex/reviewQueue.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";

// List items needing review (matches entity/review_queue.rs)
export const listPending = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", "pending"))
      .order("desc")
      .take(args.limit ?? 100);
  },
});

// List by status
export const listByStatus = query({
  args: {
    status: v.string(),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", args.status))
      .order("desc")
      .take(args.limit ?? 100);
  },
});

// Get by ID
export const get = query({
  args: { id: v.string() },
  handler: async (ctx, args) => {
    const items = await ctx.db.query("reviewQueue").collect();
    return items.find((i) => i._id.toString() === args.id) ?? null;
  },
});

// Get by raw message ID
export const getByRawMessage = query({
  args: { rawMessageId: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("reviewQueue")
      .withIndex("by_raw_message", (q) =>
        q.eq("rawMessageId", args.rawMessageId)
      )
      .collect();
  },
});

// Add item to review queue (matches entity/review_queue.rs)
export const add = mutation({
  args: {
    rawMessageId: v.string(),
    aiResult: v.any(),
    confidence: v.number(),
    reason: v.string(),
  },
  handler: async (ctx, args) => {
    // Check for existing pending review for same message
    const existing = await ctx.db
      .query("reviewQueue")
      .withIndex("by_raw_message", (q) =>
        q.eq("rawMessageId", args.rawMessageId)
      )
      .filter((q) => q.eq(q.field("status"), "pending"))
      .first();

    if (existing) {
      return existing._id;
    }

    return await ctx.db.insert("reviewQueue", {
      rawMessageId: args.rawMessageId,
      aiResult: args.aiResult,
      confidence: args.confidence,
      reason: args.reason,
      status: "pending",
      reviewedBy: undefined,
      reviewNotes: undefined,
      createdAt: Date.now(),
      reviewedAt: undefined,
    });
  },
});

// Complete review (approve)
export const complete = mutation({
  args: {
    id: v.id("reviewQueue"),
    reviewedBy: v.string(),
    notes: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      status: "approved",
      reviewedBy: args.reviewedBy,
      reviewNotes: args.notes,
      reviewedAt: Date.now(),
    });
  },
});

// Dismiss review (reject/skip)
export const dismiss = mutation({
  args: {
    id: v.id("reviewQueue"),
    reviewedBy: v.string(),
    notes: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    await ctx.db.patch(args.id, {
      status: "rejected",
      reviewedBy: args.reviewedBy,
      reviewNotes: args.notes,
      reviewedAt: Date.now(),
    });
  },
});

// Get queue stats
export const stats = query({
  args: {},
  handler: async (ctx) => {
    const pending = await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", "pending"))
      .collect();

    const approved = await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", "approved"))
      .collect();

    const rejected = await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", "rejected"))
      .collect();

    const skipped = await ctx.db
      .query("reviewQueue")
      .withIndex("by_status", (q) => q.eq("status", "skipped"))
      .collect();

    return {
      pending: pending.length,
      approved: approved.length,
      rejected: rejected.length,
      skipped: skipped.length,
    };
  },
});

File: core/crates/db/src/convex/review_queue.rs
//! Convex ReviewQueueRepository implementation

use std::sync::Arc;

use async_trait::async_trait;

use super::client::ConvexClient;
use super::error::ConvexError;
use crate::Result;
use crate::convex_args;
use crate::params::UpdateReviewStatusParams;
use crate::traits::{ReviewQueueModel, ReviewQueueRepository, ReviewQueueStats, ReviewStatus};

/// Convex-backed review queue repository
pub struct ConvexReviewQueueRepo {
    client: Arc<ConvexClient>,
}

impl ConvexReviewQueueRepo {
    pub fn new(client: Arc<ConvexClient>) -> Self {
        Self { client }
    }
}

#[async_trait]
impl ReviewQueueRepository for ConvexReviewQueueRepo {
    async fn save(&self, item: &ReviewQueueModel) -> Result<ReviewQueueModel> {
        // ReviewQueueModel: id (Uuid), raw_message_id, ai_result, confidence, reason, status, reviewed_by, review_notes, created_at, reviewed_at
        let _id: String = self
            .client
            .mutation(
                "reviewQueue:add",
                convex_args! {
                    "rawMessageId" => &item.raw_message_id,
                    "aiResult" => &item.ai_result,
                    "confidence" => item.confidence,
                    "reason" => &item.reason
                },
            )
            .await?;

        Ok(item.clone())
    }

    async fn get_by_id(&self, id: &str) -> Result<Option<ReviewQueueModel>> {
        self.client
            .query("reviewQueue:get", convex_args! { "id" => id })
            .await
    }

    async fn get_pending(&self, limit: i64, _offset: i64) -> Result<Vec<ReviewQueueModel>> {
        self.client
            .query("reviewQueue:listPending", convex_args! { "limit" => limit })
            .await
    }

    async fn get_by_status(
        &self,
        status: ReviewStatus,
        limit: i64,
        _offset: i64,
    ) -> Result<Vec<ReviewQueueModel>> {
        self.client
            .query(
                "reviewQueue:listByStatus",
                convex_args! {
                    "status" => format!("{:?}", status),
                    "limit" => limit
                },
            )
            .await
    }

    async fn update_status(
        &self,
        params: UpdateReviewStatusParams<'_>,
    ) -> Result<ReviewQueueModel> {
        // UpdateReviewStatusParams: id, status, reviewed_by, notes
        let fn_name = match params.status {
            ReviewStatus::Approved => "reviewQueue:complete",
            ReviewStatus::Rejected | ReviewStatus::Skipped => "reviewQueue:dismiss",
            _ => "reviewQueue:complete",
        };

        self.client
            .mutation_void(
                fn_name,
                convex_args! {
                    "id" => params.id,
                    "reviewedBy" => params.reviewed_by,
                    "notes" => params.notes
                },
            )
            .await?;

        self.get_by_id(params.id).await?.ok_or_else(|| {
            ConvexError::NotFound {
                entity_type: "ReviewQueue".to_string(),
                id: params.id.to_string(),
            }
            .into()
        })
    }

    async fn get_stats(&self) -> Result<ReviewQueueStats> {
        let stats: serde_json::Value = self
            .client
            .query("reviewQueue:stats", convex_args!())
            .await?;

        Ok(ReviewQueueStats {
            pending: stats["pending"].as_i64().unwrap_or(0),
            approved: stats["approved"].as_i64().unwrap_or(0),
            rejected: stats["rejected"].as_i64().unwrap_or(0),
            skipped: stats["skipped"].as_i64().unwrap_or(0),
        })
    }

    async fn count_pending(&self) -> Result<i64> {
        let stats = self.get_stats().await?;
        Ok(stats.pending)
    }

    async fn exists_for_message(&self, raw_message_id: &str) -> Result<bool> {
        let items: Vec<ReviewQueueModel> = self
            .client
            .query(
                "reviewQueue:getByRawMessage",
                convex_args! {
                    "rawMessageId" => raw_message_id
                },
            )
            .await?;
        Ok(!items.is_empty())
    }
}
